numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -341,47 +341,70 @@ format_@name@(@type@ val, npy_bool scientific,
 /**end repeat**/
 
 /*
- * over-ride repr and str of array-scalar strings and unicode to
- * remove NULL bytes and then call the corresponding functions
- * of string and unicode.
- *
- * FIXME:
- *   is this really a good idea?
- *   stop using Py_UNICODE here.
+ * Over-ride repr and str of array-scalar byte strings to remove NULL bytes and
+ * then call the corresponding functions of PyBytes_Type to generate the string
  */
 
 /**begin repeat
- * #name = string*2,unicode*2#
- * #form = (repr,str)*2#
- * #Name = String*2,Unicode*2#
- * #NAME = STRING*2,UNICODE*2#
- * #extra = AndSize*2,,#
- * #type = npy_char*2, Py_UNICODE*2#
+ * #form = repr, str#
  */
 static PyObject *
-@name@type_@form@(PyObject *self)
+stringtype_@form@(PyObject *self)
 {
-    const @type@ *dptr, *ip;
-    int len;
+    const npy_char *dptr, *ip;
+    Py_ssize_t len;
     PyObject *new;
     PyObject *ret;
 
-    ip = dptr = Py@Name@_AS_@NAME@(self);
-    len = Py@Name@_GET_SIZE(self);
-    dptr += len-1;
-    while(len > 0 && *dptr-- == 0) {
-        len--;
-    }
-    new = Py@Name@_From@Name@@extra@(ip, len);
+    ip = dptr = PyBytes_AS_STRING(self);
+    len = PyBytes_GET_SIZE(self);
+    for(dptr += len - 1; len > 0 && *dptr-- == 0; len--);
+    new = PyBytes_FromStringAndSize(ip, len);
     if (new == NULL) {
         return PyUString_FromString("");
     }
-    ret = Py@Name@_Type.tp_@form@(new);
+    ret = PyBytes_Type.tp_@form@(new);
     Py_DECREF(new);
     return ret;
 }
 /**end repeat**/
 
+/*
+ * Over-ride repr and str of array-scalar strings to remove NULL code points and
+ * then call the corresponding functions of PyUnicode_Type to generate the string
+ */
+
+/**begin repeat
+ * #form = repr, str#
+ */
+static PyObject *
+unicodetype_@form@(PyObject *self)
+{
+    Py_UCS4 *dptr, *ip;
+    Py_ssize_t len;
+    PyObject *new;
+    PyObject *ret;
+
+    if (PyUnicode_READY(self) < 0) {
+        return PyUString_FromString("");
```

Comment:
This should be `return NULL` I think

Reply:
Only if it is checked for, which I am not yet convinced of.

Reply:
It is checked for by PyObject_Repr.

As written, this code results in `SystemError: returned value with exception set` or something

Reply:
Done. Checked how other `tp_repr` and `tp_str` functions in cpython worked and they all returned NULL on failure. Also changed this for the byte strings.